File: 이분 탐색(결정알고리즘) & 그리디 알고리즘/뮤직비디오.py
import sys
# sys.stdin = open("input.txt", "r")

n, m = map(int, input().split())
dvd = list(map(int, input().split()))
maximum = max(dvd)

def count(capacity):
    cnt = 1
    tot = 0
    for x in dvd:
        if tot + x > capacity:
            cnt += 1
            tot = x
        else:
            tot += x
    return cnt

left = 1
right = sum(dvd)
result = 0
while left <= right:
    pivot = (left + right) // 2
    if pivot >= maximum and count(pivot) <= m:
        result = pivot
        right = pivot - 1
    else:
        left = pivot + 1
print(result)

File: 정렬 & 이분탐색(결정알고리즘) & 투포인트 알고리즘 & 스택/뮤직비디오.cpp
#include <stdio.h>
#include <vector>
using namespace std;

int main(void){
	// freopen("input.txt", "r", stdin);
	int n, m, i, left, right, mid, max, result, tmp, sum, cnt;
	scanf("%d%d", &n, &m);
	vector<int> arr(n);
	
	sum = 0;
	max = -2147000000;
	for(i = 0; i < n; i++){
		scanf("%d", &arr[i]);	
		sum += arr[i];
		if(arr[i] > max) max = arr[i];
	}
	
	left = 1;
	right = sum;
	while(left <= right){
		mid = (left + right) / 2;
		tmp = 0;
		cnt = 1;
		for(i = 0; i < n; i++){			
			if(tmp + arr[i] <= mid){
				tmp += arr[i];
			}
			else{
				cnt++;
				tmp = arr[i];	
			}
		}
		
		if(mid >= max && cnt <= m){
			result = mid;
			right = mid - 1;
		}
		else if(cnt > m){
			left = mid + 1;
		}
	}
	printf("%d", result);
	
	return 0;	
}
